pkg/scheduler/objects/node_iterator.go

```diff
@@ -19,103 +19,44 @@
 package objects
 
 import (
-	"math/rand"
+	"github.com/google/btree"
 )
 
 // NodeIterator iterates over a list of nodes based on the defined policy
 type NodeIterator interface {
-	// returns true if there are more values to iterate over
-	HasNext() bool
-	// returns the next node from the iterator
-	Next() *Node
-	// reset the iterator to a clean state
+	// Reset the iterator to a clean state
 	Reset()
-}
-
-// All iterators extend the base iterator
-type baseIterator struct {
-	NodeIterator
-	countIdx int
-	size     int
-	nodes    []*Node
-}
-
-// Reset the iterator to start from the beginning
-func (bi *baseIterator) Reset() {
-	bi.countIdx = 0
-}
-
-// HasNext returns true if there is a next element in the array.
-// Returns false if there are no more elements or list is empty.
-func (bi *baseIterator) HasNext() bool {
-	return !(bi.countIdx+1 > bi.size)
-}
-
-// Next returns the next element and advances to next element in array.
-// Returns nil at the end of iteration.
-func (bi *baseIterator) Next() *Node {
-	if (bi.countIdx + 1) > bi.size {
-		return nil
-	}
 
-	value := bi.nodes[bi.countIdx]
-	bi.countIdx++
-	return value
+	// ForEachNode Calls the provided function on each Node object unil it returns false
+	ForEachNode(func(*Node) bool)
 }
 
-// Default iterator, wraps the base iterator.
-// Iterates over the list from the start, position zero, to end.
-type defaultNodeIterator struct {
-	baseIterator
+type treeIterator struct {
+	accept  func(*Node) bool
+	getTree func() *btree.BTree
+	tree    *btree.BTree
 }
 
-// Create a new default iterator
-func NewDefaultNodeIterator(schedulerNodes []*Node) NodeIterator {
-	it := &defaultNodeIterator{}
-	it.nodes = schedulerNodes
-	it.size = len(schedulerNodes)
-	return it
+func (ti *treeIterator) Reset() {
```

Comment:
The Reset() method and all callers can be removed as this is a No-Op on the tree.